libraries/AP_Mount/AP_Mount_Siyi.cpp

```diff
@@ -0,0 +1,614 @@
+
+// set gimbal's lock vs follow mode
+// lock should be true if gimbal should maintain an earth-frame target
+// lock is false to follow / maintain a body-frame target
+void AP_Mount_Siyi::set_lock(bool lock)
+{
+    send_1byte_packet(SiyiCommandId::PHOTO, lock ? 3 : 4);
+}
+
+// send target pitch and yaw rates to gimbal
+// yaw_is_ef should be true if yaw_rads target is an earth frame rate, false if body_frame
+void AP_Mount_Siyi::send_target_rates(float pitch_rads, float yaw_rads, bool yaw_is_ef)
+{
+    const float pitch_rate_scalar = constrain_float(100.0 * pitch_rads / AP_MOUNT_SIYI_RATE_MAX_RADS, -100, 100);
+    const float yaw_rate_scalar = constrain_float(100.0 * yaw_rads / AP_MOUNT_SIYI_RATE_MAX_RADS, -100, 100);
+    rotate_gimbal(pitch_rate_scalar, yaw_rate_scalar, yaw_is_ef);
+}
+
+// send target pitch and yaw angles to gimbal
+// yaw_is_ef should be true if yaw_rad target is an earth frame angle, false if body_frame
+void AP_Mount_Siyi::send_target_angles(float pitch_rad, float yaw_rad, bool yaw_is_ef)
+{
+    // stop gimbal if no recent actual angles
+    uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _last_current_angle_rad_ms >= 200) {
+        rotate_gimbal(0, 0, false);
+        return;
+    }
+
+    // use simple P controller to convert pitch angle error (in radians) to a target rate scalar (-100 to +100)
+    const float pitch_err_rad = (pitch_rad - _current_angle_rad.y);
+    const float pitch_rate_scalar = constrain_float(100.0 * pitch_err_rad * AP_MOUNT_SIYI_PITCH_P / AP_MOUNT_SIYI_RATE_MAX_RADS, -100, 100);
+
+    // convert yaw angle to body-frame the use simple P controller to convert yaw angle error to a target rate scalar (-100 to +100)
+    const float yaw_bf_rad = yaw_is_ef ? wrap_PI(yaw_rad - AP::ahrs().yaw) : yaw_rad;
+    const float yaw_err_rad = (yaw_bf_rad - _current_angle_rad.z);
+    const float yaw_rate_scalar = constrain_float(100.0 * yaw_err_rad * AP_MOUNT_SIYI_YAW_P / AP_MOUNT_SIYI_RATE_MAX_RADS, -100, 100);
+
+    // rotate gimbal.  pitch_rate and yaw_rate are scalars in the range -100 ~ +100
+    rotate_gimbal(pitch_rate_scalar, yaw_rate_scalar, yaw_is_ef);
+}
+
+// take a picture.  returns true on success
+bool AP_Mount_Siyi::take_picture()
+{
+    send_1byte_packet(SiyiCommandId::PHOTO, 0);
+    return true;
+}
+
+// start or stop video recording.  returns true on success
+// set start_recording = true to start record, false to stop recording
+bool AP_Mount_Siyi::record_video(bool start_recording)
+{
+    // exit immediately if not initialised to reduce mismatch
+    // between internal and actual state of recording
+    if (!_initialised) {
+        return false;
+    }
+
+    // check desired recording state has changed
+    if (_last_record_video != start_recording) {
+        _last_record_video = start_recording;
+
+        // request recording start or stop (sadly the same message is used)
+        const uint8_t func_type = 2;
+        send_packet(SiyiCommandId::PHOTO, &func_type, 1);
```

Comment:
```suggestion
        ret = send_packet(SiyiCommandId::PHOTO, &func_type, 1);
```

Reply:
yes, record_video returns false if the send_packet fails.